azure_data_cosmos_engine/src/hash.rs

```diff
@@ -0,0 +1,537 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::fmt::Write;
+
+use crate::murmur_hash::{murmurhash3_128, murmurhash3_32};
+
+const MAX_STRING_BYTES_TO_APPEND: usize = 100;
+const MIN_INCLUSIVE_EFFECTIVE_PARTITION_KEY: &str = "";
+const MAX_EXCLUSIVE_EFFECTIVE_PARTITION_KEY: &str = "FF";
+
+/// Contains all allowed markers for component marker types.  
+mod component {
+    pub const UNDEFINED: u8 = 0x00;
+    pub const NULL: u8 = 0x01;
+    pub const BOOL_FALSE: u8 = 0x02;
+    pub const BOOL_TRUE: u8 = 0x03;
+    pub const NUMBER: u8 = 0x05;
+    pub const STRING: u8 = 0x08;
+    pub const INFINITY: u8 = 0xFF;
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum PartitionKeyValue {
+    Null,
+    Bool(bool),
+    Number(f64),
+    String(String),
+    Infinity,
+    Undefined,
+}
+
+#[derive(PartialEq, Eq)]
+pub enum PartitionKeyKind {
+    Hash,
+    MultiHash,
+}
+
+/// Returns a hex string representation of a partition key value.
+pub fn get_hashed_partition_key_string(
+    pk_value: &[PartitionKeyValue],
```

Comment:
I would have this take a `Vec` rather than a slice. In Rust, it's generally preferable to take by value as much as possible, unless that introduces `.clone()` calls. Passing by-value without `.clone`ing is generally just as fast as passing by-reference, and makes it clearer to the compiler when ownership transfers.

The caller will be done with the `PartitionKeyValue`s after they call this, so passing by value is preferable.